R/utils.R

```diff
@@ -140,3 +140,107 @@ edit.data.table = function(name, ...) {
+  negate_cols = is.call(colsub) && deparse(colsub[[1L]], 500L, backtick=FALSE) %in% c("!", "-")
+  if (negate_cols) colsub = colsub[[2L]]
+  
+  if (is.call(colsub)){
+    # fix for #1216, make sure the parentheses are peeled from expr of the form (((1:4)))
+    while(colsub %iscall% "(") colsub = as.list(colsub)[[-1L]]
+    if (length(colsub) == 3L && colsub[[1L]] == ":") {
+      if (is.name(colsub[[2L]])){
+        # cols is of the format a:c
+        rnge = chmatch(c(as.character(colsub[[2L]]), as.character(colsub[[3L]])), names(x))
+        cols = rnge[1L]:rnge[2L]
+      } else {
+        # cols is of the format 1:3
+        cols = eval(colsub)
+      }
+    } else if (colsub[[1L]] == "patterns") {
+      # each pattern gives a new filter condition, intersect the end result
+      colsub = as.list(colsub)[-1L]
+      if (length(colsub) == 1L) { #e.g. patterns("only_one_pattern")
+        cols = grep(eval(colsub[[1L]], parent.frame(2L)), names(x))
+      } else { #e.g. patterns("one", "or", "more", "patterns") 
+        x_names = names(x)
+        cols = Reduce(intersect, lapply(colsub, function(col) grep(eval(col, parent.frame(2L)), x_names)))
+      }
+    } else {
+      cols = eval(colsub, parent.frame(2L))
+    }
+  } else {
+    # if colsub is also a name in the list, we prioritize it over anything in the environment
+    if (exists(as.character(colsub), where = x)) 
+      cols = as.character(colsub) 
+    else 
+      cols = eval(colsub, parent.frame(2L), parent.frame(2L))
+  }
+                                        
+  # allow filtering via function in .SDcols, #3950
+  if (is.function(cols)) {
+    cols = lapply(x, cols)
+    if (any(idx <- vapply_1i(cols, length) > 1L | vapply_1c(cols, typeof) != 'logical' | vapply_1b(cols, anyNA)))
+      stop("When .SDcols is a function, it is applied to each column; the output of this function must be a non-missing boolean scalar signalling inclusion/exclusion of the column. However, these conditions were not met for: ", brackify(names(x)[idx]))
+    cols = unlist(cols, use.names = FALSE)
+  }
+  
+  if (origin_call == "by") {
+    #check for weird comma delimited "V1,V2,V3" and give warning message
+    #in the by eval we'll check if by = list(...) otherwise we can come here.
+  }
+  
+  if (anyNA(cols))
+    stop(origin_call, " missing at the following indices: ", brackify(which(is.na(cols))))
+  
+  if (is.logical(cols)) {
+    if (length(cols) != x_len) {
+      ## TODO change to error in 2022
+      warning("When ", origin_call, " is a logical vector, each column should have a TRUE or FALSE entry. Vector will be repeated to length of data.table. Warning will change to error in year 2022.")
+      cols = rep(cols, length.out = x_len)
+    }
+    cols = which_(cols)
+  }
+                                        
+  if (is.numeric(cols)) {
+    cols = as.integer(cols)
+    if (length(cols) == 1L && cols < 0L && abs(cols) <= x_len) cols = setdiff(seq_len(x), abs(cols))
```

Comment:
no need for `abs` call here, we already checked `cols < 0L` so just do `-cols <= -x_len`